newswires/client/src/useSearch.ts

```diff
@@ -0,0 +1,158 @@
+
+const SearchStateSchema = z.discriminatedUnion('state', [
+	z.object({
+		query: z.string(),
+		state: z.literal('loading'),
+	}),
+	z.object({
+		query: z.string(),
+		state: z.literal('error'),
+		error: z.string(),
+	}),
+	z.object({
+		query: z.string(),
+		state: z.literal('data'),
+		data: WireDataArraySchema,
+	}),
+	z.object({
+		state: z.literal('initialised'),
+	}),
+]);
+
+export type SearchState = z.infer<typeof SearchStateSchema>;
+
+const SearchHistorySchema = z.array(SearchStateSchema);
+
+function decideInitialQuery({
+	location,
+	params,
+}: {
+	location: string;
+	params?: Record<string, string>;
+}): SearchState {
+	const maybeQueryFromUrl = params?.q;
+	switch (location) {
+		case 'feed':
+			return {
+				state: 'loading',
+				query: maybeQueryFromUrl ?? '',
+			} as SearchState;
+		default:
+			return maybeQueryFromUrl
+				? ({ state: 'loading', query: maybeQueryFromUrl } as SearchState)
+				: ({ state: 'initialised' } as SearchState);
+	}
+}
+
+function fetchStoredHistory() {
+	const storedHistoryJson = localStorage.getItem('feed-searchHistory');
+
+	if (storedHistoryJson == null) return [];
+
+	try {
+		const storedHistoryObj = JSON.parse(storedHistoryJson) as unknown;
+
+		const maybeStoredHistory = SearchHistorySchema.safeParse(storedHistoryObj);
+		if (maybeStoredHistory.success) {
+			return maybeStoredHistory.data;
+		} else {
+			return [];
+		}
+	} catch (e) {
+		console.log(e);
+		return [];
+	}
+}
+
+export function useSearch() {
+	const { currentState, pushState } = useHistory();
+	const maybeSearchStateFromUrl: SearchState | undefined = decideInitialQuery({
+		location: currentState.location,
+		params: currentState.params,
+	});
+	const searchStateFromStorage = fetchStoredHistory();
+	const [searchHistory, setSearchHistory] = useState<SearchState[]>([
+		maybeSearchStateFromUrl,
+		...searchStateFromStorage,
+	]);
+
+	const searchQuery = useMemo(
+		() => currentState.params?.q,
+		[currentState.params?.q],
+	);
+
+	const pushSearchQuery = (query: string) => {
+		pushState({ location: 'feed', params: { q: query } });
+	};
+
+	const pushSearchState = useCallback(
+		(state: SearchState) => {
+			setSearchHistory((prev) => {
+				const newStack = [state, ...prev];
+				localStorage.setItem(
+					'feed-searchHistory',
+					JSON.stringify(newStack.filter((s) => s.state == 'data').slice(0, 6)),
+				);
+				return newStack.slice(0, 20);
+			});
+		},
+		[setSearchHistory],
+	);
+
+	const updateSearchQuery = (query: string) => {
+		pushSearchQuery(query);
+		pushSearchState({ query, state: 'loading' });
+	};
+
+	useEffect(() => {
+		if (searchQuery == undefined) return;
+		const quer = querify(searchQuery);
+		fetch('/api/search' + quer)
+			.then((res) => res.json())
+			.then((data) => {
+				const maybeData = WireDataArraySchema.safeParse(data);
+				if (maybeData.success) {
+					pushSearchState({
+						state: 'data',
+						query: searchQuery,
+						data: maybeData.data,
```

Comment:
discussed whether it makes sense to store all the actual search result data in the search history - fine for now, might refactor out later to make the history lighter-weight